Guides/UPI/baremetal/node-hostname-resolution.md

```diff
@@ -1,6 +1,142 @@
+          mode: 493
+          overwrite: true
+          path: /usr/local/bin/okd-resolve-node-hostname.sh
+    systemd:
+      units:
+        - name: okd-configure-hostname.service
+          enabled: true
+          contents: |
+            [Unit]
+            Description=Resolve and Set Node Hostname
+            # Removal of this file signals firstboot completion
+            ConditionPathExists=!/etc/hostname
+            # Block services relying on Networking being up.
+            Before=network-online.target
+            # Wait for NetworkManager to report its online
+            After=NetworkManager-wait-online.service
+            # Run before hostname checks
+            Before=node-valid-hostname.service
+            [Service]
+            Type=oneshot
+            RemainAfterExit=yes
+            ExecStartPre=/usr/local/bin/okd-resolve-node-hostname.sh
+            ExecStart=/bin/bash -c "hostnamectl set-hostname `cat /run/okd-node-hostname`"
+            [Install]
+            # This makes sure the systemd file is linked correctly
+            WantedBy=network-online.target
+```
+
+### Creating your `okd-resolve-node-hostname.sh`
+The script must create a file at `/run/okd-node-hostname` which is the FQDN for this node (e.g `master1.cluster.example.com`)
+
+The exact details of what this script needs to contain will vary depending on your setup. You may want to call an inventory API, perform a PTR lookup based off an IP or whatever suits your needs.
+
+This is an example `okd-resolve-node.hostname.sh` which uses the publically accessible IP of the node to perform a PTR lookup and save the result to the file.
+
+```bash
+#!/bin/bash
+
+myip=$(dig +short myip.opendns.com @208.67.222.222 -4) || myip=$(dig +short txt o-o.myaddr.l.google.com @ns1.google.com -4 | sed s/\"//g)
+myhost=$(dig +short -x ${myip} @208.67.222.222 | sed 's/\.$//' ) || myhost=$(dig +short -x ${myip} @8.8.8.8 | sed 's/\.$//')
+
+echo -n ${myhost} > /run/okd-node-hostname
+```
+
+When you have a script which works for your usecase you need to base64 encode the file and include it in the yaml as above (if you're doing this a lot look into [installer workspace](installer-workspace.md)).
+
+Finally, we need to replicate the MachineConfig for the worker nodes where relevant. If you are deploying a 3-node-cluster you can skip this step. If you are deploying a cluster with multiple or customised worker types, you will need a MachineConfig for each type.
+
+Copy the MachineConfig to a new file:
+`cp okd-configure-master-node-hostname.yaml okd-configure-worker-node-hostname.yaml`
+
+Modify the top of the file to target worker nodes and update the MachineConfig name:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: MachineConfig
+metadata:
+  labels:
+    machineconfiguration.openshift.io/role: worker
+  name: okd-configure-worker-node-hostname
+spec: ...
+```
+
+## Edit `bootstrap.ign`
```

Comment:
This is strongly not recommended. A better idea would be creating a new Ignition, which refers to unmodified `bootstrap.ign` - see `ignition.config.merge` in https://github.com/coreos/ignition/blob/master/docs/configuration-v3_2.md

Reply:
I agree to edit the `bootstrap.ign` is not ideal but to use the merge methodology would be troublesome, I would either need to base64 the vanilla `bootstrap.ign` and include that, or it would require the files to be accessible on a server - which doesn't occur in my deploy pattern.

In my deploy usecase, the physical machines are net-booted into a "rescue" OS. The ignition files are sftp-ed into the rescue OS and then we use the [run from container](https://coreos.github.io/coreos-installer/getting-started/#run-from-a-container) methodology to install to the local disk.

In future we will likely switch to embedding the ignition into an ISO created for each server. 

I can switch the docs to base64 encode the original in but I actually think this adds more room for human error + vastly reduces the ability for someone to inspect/debug what's going on the system when something goes wrong.

Reply:
Oh, hmm, okay, lets describe this generally this is not recommended - but still possible of course

Reply:
Yeah, unfortunately, there's no config for us to modify to impact the rendering of the bootstrap.ign in earlier steps to any significant degree which would be the far superior way.

You're really not going to like the workaround for nodes which have dual IPs (public/private) which requires a custom bootkube.sh to ensure cluster traffic uses the correct interfaces!

Reply:
I've added text that mentions editing the rendered filed are not recommended and pointed them to the merge functionality if that's something which suits their flow.